ts/input/tex/ams/AmsItems.ts

```diff
@@ -117,3 +118,105 @@ export class MultlineItem extends ArrayItem {
     this.factory.configuration.tags.end();
   }
 }
+
+/**
+ * StackItem for handling flalign, xalignat, and xxalignat environments.
+ */
+export class FlalignItem extends EqnArrayItem {
+
+  /**
+   * @override
+   */
+  get kind() {
+    return 'flalign';
+  }
+
+
+  /**
+   * @override
+   */
+  constructor(factory: any, public name: string, public numbered: boolean,
+              public padded: boolean, public center: boolean) {
+    super(factory);
+    this.factory.configuration.tags.start(name, numbered, numbered);
+  }
+
+  /**
+   * @override
+   */
+  public EndEntry() {
+    super.EndEntry();
+    const n = this.getProperty('xalignat') as number;
+    if (!n) return;
+    if (this.row.length > n) {
+      throw new TexError('XalignOverflow', 'Extra %1 in row of %2', '&', this.name);
+    }
+  }
+
+
+  /**
+   * @override
+   */
+  public EndRow() {
+    let cell: MmlNode;
+    let row = this.row;
+    //
+    //  For xalignat and xxalignat, pad the row to the expected number if it is too short.
+    //
+    const n = this.getProperty('xalignat') as number;
+    while (row.length < n) {
+      row.push(this.create('node', 'mtd'));
+    }
+    //
+    //  Insert padding cells between pairs of entries, as needed for "fit" columns,
+    //    and include initial and end cells if that is needed.
+    //
+    this.row = [];
+    if (this.padded) {
+      this.row.push(this.create('node', 'mtd'));
+    }
+    while ((cell = row.shift())) {
+      this.row.push(cell);
+      cell = row.shift();
+      if (cell) this.row.push(cell);
+      if (row.length || this.padded) {
+        this.row.push(this.create('node', 'mtd'));
+      }
+    }
+    //
+    if (this.row.length > this.maxrow) this.maxrow = this.row.length;
+    super.EndRow();
+    //
+    // For full-width environments with labels that aren't supposed to take up space,
+    //   move the label into a zero-width mpadded element that laps in the proper direction.
+    //
+    const mtr = this.table[this.table.length - 1];
+    if (this.getProperty('zeroWidthLabel') && mtr.isKind('mlabeledtr')) {
+      const mtd = NodeUtil.getChildren(mtr)[0];
+      const side = this.factory.configuration.options['tagSide'];
+      const def = {width: 0, ...(side === 'right' ? {lspace: '-1width'} : {})};
```

Comment:
Yes.  That says shift to the left by the width of the contents (the 1 is necessary according to the spec).  This means that when the label is on the right (`side === 'right'`), then make the label 0 width by shifted to the left by its width.  That means that when the table is full width (100%), the label will overlap the 100% width rather than extend past the 100% width (making for a wider element than full width).